include/mbgl/style/layer.hpp

```diff
@@ -98,20 +91,30 @@ class Layer : public mbgl::util::noncopyable {
         }
     }
 
+    LayerType getType() const;
     const std::string& getID() const;
 
     // Visibility
     VisibilityType getVisibility() const;
-    void setVisibility(VisibilityType);
+    virtual void setVisibility(VisibilityType) = 0;
 
     // Zoom range
     float getMinZoom() const;
-    void setMinZoom(float) const;
     float getMaxZoom() const;
-    void setMaxZoom(float) const;
+    virtual void setMinZoom(float) = 0;
+    virtual void setMaxZoom(float) = 0;
 
     // Private implementation
-    const std::unique_ptr<Impl> baseImpl;
+    class Impl;
+    Immutable<Impl> baseImpl;
```

Comment:
`Style` at a minimum would need to be friended. `GeometryTile` also uses it; it's possible that use could be eliminated. I'd like to leave this for a followup change though.